### PR TITLE
std::result_of removed in C++20

### PR DESCRIPTION
**Pull request by maxgolov**
Fix for #104 : when OpenTelemetry is compiled with latest update of Visual Studio 2019 in C++17 mode - it'd generate a warning, and in C++20 it won't compile because std::result_of has been removed in C++20. Adjust the header to stop using a compiler feature that's been removed from standard.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -20,6 +20,11 @@ include(CTest)
 
 find_package(Threads)
 
+if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+  # Options for Visual C++ compiler
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zc:__cplusplus")
```
Comment (g-easy) [inline CMakeLists.txt:25]: Could you please describe in the comment that Zc does?

**Comment by g-easy** [comment]
Failing test:
```
[ RUN      ] RandomTest.GenerateRandomBuffer
sdk/test/common/random_test.cc:30: Failure
Value of: std::equal(std::begin(buf1), std::end(buf1), std::begin(buf2))
  Actual: true
Expected: false
[  FAILED  ] RandomTest.GenerateRandomBuffer (0 ms)
```

cc @pyohannes

File: CMakeLists.txt (hunk 1 above)
Comment (maxgolov) [reply CMakeLists.txt:25]: > Failing test:
> [  FAILED  ] RandomTest.GenerateRandomBuffer (0 ms)

Unrelated to my commit though. It passed initially, failed after I merged the latest change.

Comment (maxgolov) [reply CMakeLists.txt:27]: > Could you please describe in the comment that Zc does?

Done. Added a very short summary of what's written here:
https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=vs-2019

Comment (g-easy) [reply CMakeLists.txt:27]: Nice! Thank you!

**Comment by reyang** [comment]
@maxgolov please rebase, thanks.

**Pull request merged**